proxy/coco/src/state.rs

```diff
@@ -0,0 +1,733 @@
+use std::path::{self, PathBuf};
+use std::sync::Arc;
+use std::time::Duration;
+
+use tokio::sync::Mutex;
+
+use librad::git::local::{transport, url::LocalUrl};
+use librad::git::refs::Refs;
+use librad::git::storage;
+use librad::keys;
+use librad::meta::entity;
+use librad::meta::project as librad_project;
+use librad::meta::user;
+use librad::net::gossip::PeerInfo;
+use librad::net::peer::PeerApi;
+use librad::paths;
+use librad::peer::PeerId;
+use librad::uri::{RadUrl, RadUrn};
+use radicle_keystore::sign::Signer as _;
+use radicle_surf::vcs::git;
+use radicle_surf::vcs::git::git2;
+
+use crate::error::Error;
+use crate::project;
+use crate::seed::Seed;
+use crate::signer;
+use crate::user::{verify as verify_user, User};
+
+/// Thread-safe wrapper for [`State`].
+pub type Lock = Arc<Mutex<State>>;
+
+impl From<State> for Lock {
+    fn from(state: State) -> Self {
+        Arc::new(Mutex::new(state))
+    }
+}
+
+/// High-level interface to the coco monorepo and gossip layer.
+pub struct State {
+    /// Internal handle on [`PeerApi`].
+    pub(crate) api: PeerApi<keys::SecretKey>,
+}
+
+impl State {
+    /// Create a new [`State`] given a [`PeerApi`].
+    #[must_use]
+    pub fn new(api: PeerApi<keys::SecretKey>, signer: signer::BoxedSigner) -> Self {
+        let paths = api.paths();
+
+        // Turns out we don't need to call the git commands from the command line.
+        // All we need is to register the git transport and this is the same
+        // thing. This means we can use git2 to execute actions. So instead of
+        // implementing the checkout via git clone as a Command, we use
+        // [`git2::RepoBuilder`].
+        transport::register(transport::Settings {
+            paths: paths.clone(),
+            signer,
+        });
+
+        Self { api }
+    }
+
+    /// Returns the [`PathBuf`] to the underlying monorepo.
+    #[must_use]
+    pub fn monorepo(&self) -> PathBuf {
+        self.api.paths().git_dir().join("")
+    }
+
+    /// Returns the underlying [`paths::Paths`].
+    #[must_use]
+    pub fn paths(&self) -> paths::Paths {
+        self.api.paths().clone()
+    }
+
+    /// Convenience method to trigger a reopen of the storage.
+    ///
+    /// # Errors
+    ///
+    /// When the underlying lock acquisition fails or opening the storage.
+    pub fn reopen(&self) -> Result<(), Error> {
+        self.api.storage().reopen()?;
+        Ok(())
+    }
+
+    /// Check the storage to see if we have the given commit for project at `urn`.
+    ///
+    /// # Errors
+    ///
+    ///   * Checking the storage for the commit fails.
+    ///
+    /// # Panics
+    ///
+    ///   * Unable to acquire the lock.
+    pub fn has_commit(&self, urn: &RadUrn, oid: impl Into<git2::Oid>) -> Result<bool, Error> {
+        Ok(self.api.storage().has_commit(urn, oid.into())?)
+    }
+
+    /// Our current peers [`PeerId`].
+    #[must_use]
+    pub fn peer_id(&self) -> PeerId {
+        self.api.peer_id().clone()
```

Comment:
#890

Reply:
> the peer_id is not a stateful matter

It is, actually, because everything below `refs/<namespace>/rad/*` is expected to be signed by the `PeerId` which was advertised during the handshake. Iow, one cannot simply change the local key without rewriting the repository, or else bad things will happen (although we haven't yet enumerated all of them).